components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/DefaultTopSitesStorage.kt

```diff
@@ -0,0 +1,93 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.top.sites
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import mozilla.components.browser.storage.sync.PlacesHistoryStorage
+import mozilla.components.feature.top.sites.ext.toTopSite
+import mozilla.components.feature.top.sites.TopSite.Type.FRECENT
+import mozilla.components.feature.top.sites.ext.hasUrl
+import mozilla.components.support.base.observer.Observable
+import mozilla.components.support.base.observer.ObserverRegistry
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * Default implementation of [TopSitesStorage].
+ *
+ * @param pinnedSitesStorage An instance of [PinnedSiteStorage], used for storing pinned sites.
+ * @param historyStorage An instance of [PlacesHistoryStorage], used for retrieving top frecent
+ * sites from history.
+ * @param defaultTopSites A list containing a title to url pair of default top sites to be added
+ * to the [PinnedSiteStorage].
+ */
+class DefaultTopSitesStorage(
+    private val pinnedSitesStorage: PinnedSiteStorage,
+    private val historyStorage: PlacesHistoryStorage,
+    private val defaultTopSites: List<Pair<String, String>> = listOf(),
+    coroutineContext: CoroutineContext = Dispatchers.IO
+) : TopSitesStorage, Observable<TopSitesStorage.Observer> by ObserverRegistry() {
+
+    private var scope = CoroutineScope(coroutineContext)
+
+    // Cache of the last retrieved top sites
+    var cachedTopSites = listOf<TopSite>()
```

Comment:
I'm not convinced we need to add this to the storage layer. We should probably be refactoring our usages in the apps to use `getTopSites` correctly (and use the app's caching mechanism) rather than leaking that into the component.

Reply:
I don't want to hold up this review any longer so let's go ahead and fix this in a follow-up.